packages/ckeditor5-coremedia-blocklist/src/blocklistui.ts

```diff
@@ -0,0 +1,443 @@
+    // Focus the blocklist view if the balloon is visible and the Tab key has been pressed.
+    this.editor.keystrokes.set(
+      "Tab",
+      (data, cancel) => {
+        if (this.#isBlocklistVisible() && !this.blocklistActionsView.focusTracker.isFocused) {
+          this.blocklistActionsView.focus();
+          cancel();
+        }
+      },
+      {
+        // Use the normal priority because the link UI navigation is more important
+        priority: "normal",
+      }
+    );
+
+    // Close the panel on the Esc key press when the editable has focus and the balloon is visible.
+    this.editor.keystrokes.set("Esc", (data, cancel) => {
+      if (this.#isBlocklistVisible()) {
+        this.#hideBlocklistBalloon();
+        cancel();
+      }
+    });
+
+    // Close on click outside of balloon panel element.
+    clickOutsideHandler({
+      emitter: this.blocklistActionsView,
+      activator: () => this.#isBlocklistViewInBalloonPanel(),
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      contextElements: () => [this.#balloon!.view.element!],
+      callback: () => this.#hideBlocklistBalloon(),
+    });
+  }
+
+  /**
+   * Returns `true` when {@link #blocklistActionsView} is in the balloon, and it is
+   * currently visible.
+   */
+  #isBlocklistVisible(): boolean {
+    if (!this.#balloon) {
+      return false;
+    }
+    return !!this.blocklistActionsView && this.#balloon.visibleView === this.blocklistActionsView;
+  }
+
+  /**
+   * Returns `true` when {@link #blocklistActionsView} is in the balloon.
+   */
+  #isBlocklistViewInBalloonPanel(): boolean {
+    if (!this.#balloon) {
+      return false;
+    }
+    return !!this.blocklistActionsView && this.#balloon.hasView(this.blocklistActionsView);
+  }
+
+  /**
+   * Returns a list of blocked words, based on the current cursor position.
+   * A list will only be returned if the selection is collapsed.
+   *
+   * @returns All words, displayed by markers at the current position
+   */
+  #getSelectedBlocklistWords(): string[] | undefined {
+    const view = this.editor.editing.view;
+    const selection = view.document.selection;
+
+    const firstPosition = selection.getFirstPosition();
+    if (!firstPosition) {
+      return undefined;
+    }
+
+    if (selection.isCollapsed) {
+      return this.#getAllBlocklistedWordsForPosition(firstPosition);
+    }
+    return undefined;
+  }
+
+  /**
+   * Returns the raw text in the currently selected editor content.
+   *
+   * @private
+   */
+  #getSelectedText(): string {
+    const view = this.editor.editing.view;
+    const selection = view.document.selection;
+    const range = selection.getFirstRange();
+
+    const isTextProxy = (node: unknown): node is TextProxy =>
+      // @ts-expect-error weifhwe
+      node.data !== undefined;
+
+    // Regex to remove special characters, like commas, that appear when a highlighted section is within
+    // the selection
+    const regex = /[^a-zA-Z0-9\s-]/g;
```

Comment:
I would suggest to extract the string replacement in an extra method and to write a test.

Regex are always a bit hard to read and a test helps a lot. Furthermore, we can prove that it really does what we expect.